src/field.cpp

```diff
@@ -461,6 +461,30 @@ field_id field_from_ident(const std::string &field_ident)
     return fd_null;
 }
 
+void map::create_burnproducts( const tripoint p, item &fuel ) {
+	std::vector<material_id> all_mats = fuel.made_of();
+	if( !all_mats.empty() )
+	{
+	    units::mass fuel_weight = fuel.weight( false );
+	    //Items that are multiple materials are assumed to be equal parts each.
+	    units::mass by_weight = fuel_weight / all_mats.size();
+	    for( auto &mat : all_mats ) {
+	        if( !mat->burn_products().empty() ) {
+	            mat_burn_products all_bp = mat->burn_products();
```

Comment:
`burn_products` is called twice (here and above), calling it before the `if` clause and using `all_bp` in the condition would be more efficient. But you don't need this variable (and the emptiness check) at all. It's perfectly valid to iterate over an empty container, and it's valid to iterator over a function result:
```C++
for( auto &bp : mat->burn_products() ) {
```